discovery/discovery_test.go

```diff
@@ -47,6 +59,57 @@ func StubOrchestrators(addresses []string) []*lpTypes.Transcoder {
 	return orchestrators
 }
 
+type stubOrchestratorServer struct {
+	first bool
+}
+
+func (s *stubOrchestratorServer) GetOrchestrator(context context.Context, req *net.OrchestratorRequest) (*net.OrchestratorInfo, error) {
+	if s.first {
+		time.Sleep(100 * time.Millisecond)
+		s.first = false
+	}
+	return &net.OrchestratorInfo{Transcoder: "transcoderfromtestserver"}, nil
+}
+
+func (s *stubOrchestratorServer) Ping(context context.Context, req *net.PingPong) (*net.PingPong, error) {
+	return nil, nil
+}
+
+func initOrchServer() *bufconn.Listener {
+	lis = bufconn.Listen(bufSize)
+	s := grpc.NewServer()
+	net.RegisterOrchestratorServer(s, &stubOrchestratorServer{first: true})
+	go func() {
+		if err := s.Serve(lis); err != nil {
+			log.Fatalf("Server exited with error: %v", err)
+		}
+	}()
+	return lis
+}
+
+func testDialer(string, time.Duration) (gnet.Conn, error) {
+	return lis.Dial()
+}
+
+func TestDeadLock(t *testing.T) {
```

Comment:
Maybe we could stub out the `server.GetOrchestrator` part rather than building out a bunch of scaffolding around a real network etc? Something like:

```golang
type orchestratorPool struct {
 ...
orchInfo: func(context.Context, *url.URL) (*net.OrchestratorInfo, error)
}

func NewOrchestratorPool(...) *orchestratorPool {
...
// bcast is only used for server.GetOrchestratorInfo so can prob be removed from the struct at this point
bcast := core.NewBroadcaster(node)
orchInfo := func(context.Context, *url.URL) (*net.OrchestratorInfo, error) { return server.GetOrchestratorInfo(ctx, bcast,url)
return &orchestratorPool{orchInfo}
}
...
// testing code
pool := NewOrchestratorPool(nil, addresses)
pool.orchInfo = func ... // avoids the whole network dance!
```

@angyangie and I have encountered similar issues a couple times in the past and came close to implementing something like this a few times, but have always managed to work around it... might be unavoidable now.

What do you think?

Reply:
I prefer not to have dynamic code - it is harder to follow. But it is unavoidable of course, and should be done for new code.
For that particular unit test I would prefer to left it as is - it shows example how to test `grpc` services/mock `grpc` service's server side. 

For the future - maybe it's better not to change `orchestratorPool`, but change `server.GetOrchestratorInfo` to `server.globalVar.GetOrchestratorInfo` - that way we will be able to replace `server.globalVar` for testing - this will give us ability to mock it for all the code that uses `server.GetOrchestratorInfo`?

Reply:
> For that particular unit test I would prefer to left it as is

Disagree. gRPC is not required here, and really clutters up the testing; it's difficult to distinguish what's essential to the test and what's (avoidable) scaffolding.

Agree that using function pointers is a hackish, but it seems to be the least worst option.

> For the future - maybe it's better not to change orchestratorPool, but change server.GetOrchestratorInfo to server.globalVar.GetOrchestratorInfo

If you really want to use a global rather than a member var, the global can be within `discovery.go` : `var getOrchInfo = server.getOrchestratorInfo`. Within the testing procedure, we can reset `getOrchInfo = <stub>`.

Reply:
Made the change